# DataFrames

DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [1]:
import numpy as np
import pandas as pd

In [29]:
# help(pd.DataFrame)

### Creating a DataFrame from Python Objects

In [2]:
np.random.seed(101)
mydata = np.random.randint(0,101,(6,4))

In [3]:
mydata

array([[95, 11, 81, 70],
       [63, 87, 75,  9],
       [77, 40,  4, 63],
       [40, 60, 92, 64],
       [ 5, 12, 93, 40],
       [49, 83,  8, 29]])

In [28]:
myindex = ['Casa','Khou','Rabat','Media','Kech','Tang']

In [29]:
mycolumns = ['Jan','Feb','Mar','Sept']

In [30]:
pd.DataFrame(data=mydata)

0   1   2   3
0  95  11  81  70
1  63  87  75   9
2  77  40   4  63
3  40  60  92  64
4   5  12  93  40
5  49  83   8  29

In [31]:
pd.DataFrame(data=mydata,index=myindex)

0   1   2   3
Casa   95  11  81  70
Khou   63  87  75   9
Rabat  77  40   4  63
Media  40  60  92  64
Kech    5  12  93  40
Tang   49  83   8  29

In [32]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df

Jan  Feb  Mar  Sept
Casa    95   11   81    70
Khou    63   87   75     9
Rabat   77   40    4    63
Media   40   60   92    64
Kech     5   12   93    40
Tang    49   83    8    29

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6 entries, Casa to Tang
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   0       6 non-null      int32
 1   1       6 non-null      int32
 2   2       6 non-null      int32
 3   3       6 non-null      int32
dtypes: int32(4)
memory usage: 316.0+ bytes


## Reading a .csv file for a DataFrame

### CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

#### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")## CSV

#### Print your current directory file path with pwd

In [41]:
#pwd

#### List the files in your current directory with ls

In [42]:
#ls

In [43]:
df = pd.read_csv('assets/tips.csv')

In [44]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

## DataFrames

### Obtaining Basic Information About DataFrame

In [45]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [46]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [48]:
df.head(3)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458

In [49]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [51]:
len(df)

244

In [52]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [53]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

### Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

#### COLUMNS

We will begin be learning how to extract information based on the columns

In [54]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [63]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [60]:
type(df['total_bill'])

pandas.core.series.Series

#### Grab Multiple Columns

In [61]:
# Note how its a python list of column names! Thus the double brackets.
df[['total_bill','tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [89]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [90]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [81]:
df['price_per_person'] = df['total_bill'] / df['size']

In [82]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [83]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Adjust Existing Columns

In [84]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [85]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Remove Columns

In [86]:
# df.drop('tip_percentage',axis=1)

In [91]:
#remove columns permanently
df = df.drop("tip_percentage",axis=1)

In [92]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [93]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [95]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [98]:
#df.set_index('Payment ID')
df = df.set_index('Payment ID')

In [99]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [100]:
df = df.reset_index()

In [101]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

### ROWS

Let's now explore these same concepts but with Rows.

In [103]:
df = df.set_index('Payment ID')

In [104]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

#### Grab a Single Row

In [109]:
# Integer Based
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [ ]:
# key error df[0] !!

In [106]:
# Name Based
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [141]:
type(df.loc['Sun2959'])

pandas.core.series.Series

In [111]:
# key error df['Sun2959'] !!

#### Grab Multiple Rows

In [112]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [113]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun5260                11.84    Nathaniel Harris  4676137647685994

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values.

In [114]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [115]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

In [116]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [118]:
one_row = df.iloc[0]
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [119]:
type(one_row)

pandas.core.series.Series

In [128]:
df.tail()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sat1766          27.18  2.00  Female    Yes  Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes  Sat  Dinner     2   
Sat17            17.82  1.75    Male     No  Sat  Dinner     2   
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Sun2959                 8.49  Christy Cunningham  3560325168603410

In [138]:
#df = df.append(one_row).tail() 

In [137]:
df = pd.concat([df,one_row.to_frame().T])

In [139]:
df.loc['test'] = [16.99, 1.01, 'Female', 'No', 'Sun', 'Dinner', 2, 8.49, 'Christy Cunningham', 3560325168603410]

In [140]:
df.tail()

total_bill   tip     sex smoker  day    time size price_per_person  \
Sat1766      27.18   2.0  Female    Yes  Sat  Dinner    2            13.59   
Sat3880      22.67   2.0    Male    Yes  Sat  Dinner    2            11.34   
Sat17        17.82  1.75    Male     No  Sat  Dinner    2             8.91   
Sun2959      16.99  1.01  Female     No  Sun  Dinner    2             8.49   
test         16.99  1.01  Female     No  Sun  Dinner    2             8.49   

                 Payer Name         CC Number  
Sat1766      Monica Sanders  3506806155565404  
Sat3880          Keith Wong  6011891618747196  
Sat17          Dennis Dixon     4375220550950  
Sun2959  Christy Cunningham  3560325168603410  
test     Christy Cunningham  3560325168603410